### DISCUSSION TITLE
Create namespaces for lanes

### DISCUSSION THREAD

**Discussion by AliSoftware** [fastlane/fastlane#18164] [Ideas]
As our `Fastfile` grows quickly with a lot of lanes, we would love to be able to group some lanes together in some kind of namespace to make both their naming, their discoverability, and the documentation easier to manage.

The idea would be pretty similar to Rake's `namespace` – which allows you to group rake `tasks` together using `namespace :foo do … end`. 

## Why ?

This feature would especially be very useful when you have many lanes in your `Fastfile`, many of them related. For example, in [WordPress-iOS](https://github.com/wordpress-mobile/WordPress-iOS/blob/19c662ce16146959b7f2556e1d67231473a9fa66/Scripts/fastlane/Fastfile), we have:

 - Multiple lanes to do various tasks around building, testing, … – which are usually called by our devs
 - Multiple lanes around release management, especially cutting a new release branch and bumping the version, or creating a new beta during our consolidation phase before release, etc – which are usually called by the Release Managers (me)
 - Multiple lanes used for CI – publish to testflight, etc – which are rarely called by our users, but only by CircleCI / GitHub Actions workflows
 - Multiple lanes related to onboarding and ADC (do all the necessary steps to onboard a new joiner like adding them to the relevant teams and all; add a new device to the portal when someone has a new test device; update the profiles for all our white label variations and add all devices to profiles…)

That ends up being a lot of lanes, for which we either need to provide very long names with ugly underscores just to make it clear what they are about, or find some way to make it clear who and what is the lane for, but also makes `fastlane lanes` and the documentation about all the available lanes… quite a mess, as all those lanes ends up being listed in an arbitrary order that doesn't match our expected categorisation and grouping at all

### DSL

In the `Fastfile` DSL, I'm suggesting something like this:

```ruby
# Lanes for the Release Manager
namespace :release do
  namespace :cut do
    lane :start do
      # Actions to start a cut of a new release branch when we're ready to freeze a version for beta
    end
    lane :finish
      # Actions to finish the cut of the new release branch once we've done all the manual adjustments (and the initial release cut PR is approved and merged)
    end
  end
  lane :submit do
    # Actions to submit the release to the AppStore for review
  end
  lane :publish do
    # Actions to publish the release live once approved
  end
end

# Lanes to build and test the code
platform :ios do
  namespace :code do
    lane :build do … end
    lane :test do … end
    …
  end
end

# Lanes for onboarding
namespace :configure do
  lane :device do … end # To register a new device's UUID to the portal and regenerate everything needed
  lane :profiles do … end # To update all the provisioning profiles
  lane :secrets do … end # To configure our secret private repository containing all the necessary API Keys and secrets
  …
end
```

### Referencing lane names in CLI

In fact a similar idea to namespacing has been proposed and started to be implemented 2 years ago in (https://github.com/fastlane/fastlane/pull/11695) but got abandoned, because using `:` caused conflicts with the parsing of named parameters passed to the CLI (iow, should `fastlane foo bar:baz` be interpreted as "platform foo, lane baz in namespace bar" or as "lane foo with parameter bar having value baz"?).

That mean that we might need a different syntax to reference namespace lanes than the classic `ns1:ns2:task` syntax used by `Rake`, in order to not conflict with the existing fastlane syntax for parameters.
1. Maybe using `::` for namespacing (similar to how Ruby modules are referenced)?
2. Or maybe requiring that every namespaced qualified name is referenced with a leading `:` to disambiguate with lane parameters?
3. Or use `/`?
4. Or… still use `:` like Rake, and find a way of making it work without conflicting with parameters passed in the CLI.
    - For example maybe we could say that when referencing a namespaced lane, we should never provide the platform as a separate parameter but instead consider the platform like a namespace itself, leading to `fastlane ios:code:build` as `lane :build` in `namespace :code` is all nested inside `platform :ios` in the `Fastfile`.
    - This means that the lane name will in that case always expected to be the first parameter
    - That might actually work quite well given that the platform is kind of a namespace already, to group lanes by platform, and also because some lanes don't really belong to a specific platform (e.g. our lanes to cut a release branch and call the GitHub API to prepare the release work… are not really specific to iOS or Android) so for those it would make sense to not be inside a given `platform :xx do…end` block to begin with.

```sh
## Syntax Suggestion 1: double '::'
fastlane ios code::build config:release # explicit mention of the platform name, with parameters passed to the lane
fastlane code::build config:release # implicit platform
fastlane release::cut::finish version:5.6 # no platform in CLI args
fastlane release::submit # no params

## Syntax Suggestion 2: leading ':'
fastlane ios :code:build config:release # explicit mention of the platform name, with parameters passed to the lane
fastlane :code:build config:release # implicit platform
fastlane :release:cut:finish version:5.6 # no platform in CLI args
fastlane :release:submit # no params

## Syntax Suggestion 3: using '/'
fastlane ios code/build config:release # explicit mention of the platform name, with parameters passed to the lane
fastlane code/build config:release # implicit platform
fastlane release/cut/finish version:5.6 # no platform in CLI args
fastlane release/submit # no params

## Syntax Suggestion 4: use ':', but first param is always the lane name, prefixed with `platform:` if needed
fastlane ios:code:build config:release # explicit mention of the platform name, with parameters passed to the lane
fastlane code:build config:release # implicit platform
fastlane release:cut:start version:5.6 # no platform in CLI args
fastlane release:cut:finish version:5.6 # implicit :ios platform
fastlane release:submit
```

Any other suggestions welcome.

\cc @joshdholtz

**Comment by AliSoftware** [comment]
Another neat benefit of allowing namespaces is that this might also enable us to allow lanes to have the same name as an existing action, as long as the lane is in a namespace.

For example, many projects have a lane to submit to TestFlight. But we can't (at least warnings in fastlane highly discourage us to) name the lane just `testflight` — despite that name being very apt for such a task — because it conflicts with the official `testflight` action name and might be ambiguous.

But if we were to name a `lane :testflight`… inside a `namespace :adc` for example, then the fully qualified name `adc::testflight` (or whatever syntax we end up choosing for that) would _not_ conflict anymore!

_(Naming things is already one of the hardest problem in computer science, so yay for nicer, more understandable lane names!)_

**Reply by joshdholtz** [reply]
I run into this all the time 😛 I hope we can actually make this work!

**Reply by AliSoftware** [reply]
I edited the description a bit to suggest more possible syntaxes for referencing the lane names in CLI.

Personally I think my preferred one is option 4. Then, if not possible, then use option 2.

All of those suggestions might also allow us to reference lanes by their unqualified, un-namespaced name as a shorthand if there's no ambiguity in the lane name resolution – similar to how we do lane resolution in a given `platform` if it's not explicitly provided – and use the fully-qualified name if we want or need to be explicit.

**Comment by AliSoftware** [comment]
An alternative name for the `namespace` keyword in the DSL: `group`

```ruby
group :release do
  group :cut do
    lane :start do
      # …
    end
  end
end
```

**Reply by joshdholtz** [reply]
I think I like `group` the best! The shortness of it looks a little bit cleaner to read 😊

**Reply by mokagio** [reply]
`group` and `namespace` could be aliases. Folks used to Rake might prefer `namespace`, while people who like succinctness can use `group`.

Or maybe it's easier to provide only one option and avoid the "what's the difference..." FAQ 🤔

**Comment by ffittschen** [comment]
This is a great suggestion! We also ran into this recently, and ended up with extracting parts of our original `Fastfile` into a `BuildFastfile`, `TestsFastfile` and `ToolingFastfile`, which are then imported in the normal `Fastfile`. Namespaces / groups would have been a nice addition to making the lane names more concise, and the ability to use names taken by actions is a nice benefit as well.

From these suggestions, I'd also prefer `group` + option 4.
